### ISSUE TITLE
The imperative programming paradigm

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#7464]
```python
r = fluid.data.mnist.train()

W = fluid.Tensor()
b = fluid.Tensor()

with fluid.While(iters=100):
    mb = r.next()

    x = fluid.layer.data(mb.image)
    l = fluid.layer.data(mb.label)
    y = fluid.layer.fc(x, W, b)
    cost = fluid.layer.mse(y, l)

    fluid.optimize(cost)
    
fluid.print(W, b)
```

**Comment by wangkuiyi** [comment]
```python
r = fluid.data.mnist.train()

W = fluid.Tensor()
b = fluid.Tensor()

with fluid.While(iters=100):
    mb = r.next()

    x = fluid.layer.data(mb.image)
    l = fluid.layer.data(mb.label)
    y = fluid.layer.fc(x, W, b)
    cost = fluid.layer.mse(y, l)

    fluid.optimize(cost)
    
fluid.print(W, b)

mb = r.next()[1:2]
x = fluid.layer.data(mb)
y = fluid.layer.fc(x, W, b)
fluid.print(y)
```

**Comment by wangkuiyi** [comment]
We can use [atexit](https://docs.python.org/2/library/atexit.html) to register a *launcher*.

For example, we could have a launcher which saves the ProgramDesc into a file, or another launcher that calls the C++ `Executor` class to run the ProgramDesc, or yet another launcher that calls a transpiler (or a pipe of transpilers) to convert the ProgramDesc into C++ programs and build them into a binary executable.

**Comment by wangkuiyi** [comment]
Do we still need the data layer, or should we just call `fluid.layer.fc(input=mb.image, ...)`?

**Comment by reyoung** [comment]
![default](https://user-images.githubusercontent.com/728699/34864292-8f2eb192-f7ae-11e7-82dd-fbb99ae094b6.png)

For fully parallel, we might use `goroutine` and `channel`.

**Comment by wangkuiyi** [comment]
We might be able to use Python multiprocess to communicate between a Python process and a Fluid process. -- from @emailweixu

**Comment by typhoonzero** [comment]
I took a glance at http://ray.readthedocs.io/en/latest/tutorial.html, for distributed training fluid API, we may have two main design:

1. Using channel
```python
grad_param_mapping = dict()

def def_trainer():
    r = fluid.data.mnist.train()
    W = fluid.Tensor()
    b = fluid.Tensor()
    with fluid.While(iters=100):
        mb = r.next()

        x = fluid.layer.data(mb.image)
        l = fluid.layer.data(mb.label)
        y = fluid.layer.fc(x, W, b)
        cost = fluid.layer.mse(y, l)
        grads, params = fluid.bp(cost)
        for idx, grad in enumerate(grads):
        grad_param_mapping[grad.name] = params[idx]
        fluid.send(grads, overwrite=params)

def def_pserver():
    with fluid.listen_and_serv() as server:
        with fluid.While():
            grad = server.recv_chan.get()
            param = grad_param_mapping.get(grad.name)
            fluid.optimize(param, grad)
            server.return_chan.push(param)
```
2. Using actor mode
```python
@fluid.remote
class Pserver:
    def __init__(self, name_map):
        self.grad_param_mapping = name_map
    def send(self, grad):
        param = self.grad_param_mapping.get(grad.name)
        fluid.optimize(param, grad)
    def get(self, varname):
        return self.grad_param_mapping.get(grad.name)

@fluid.remote
def trainer():
    ...

```

**Comment by QiJune** [comment]
A simple demo of integrating fluid with ray.
```python
import ray
import paddle.v2 as paddle
import paddle.v2.fluid as fluid
import numpy

ray.init()

NUM_WORKERS = 2
BATCH_SIZE = 128


def mnist_network():
    img = fluid.layers.data(name='img', shape=[784])
    hidden = fluid.layers.fc(img, size=100, act='tanh',
                             param_attr='fc.w',
                             bias_attr='fc.b')
    prediction = fluid.layers.fc(hidden, size=10, act='softmax',
                                 param_attr='sftmax.w',
                                 bias_attr='sftmax.b')
    label = fluid.layers.data(name='label',shape=[1],
                              dtype='int64')
    loss = fluid.layers.cross_entropy(input=prediction, label=label)


    avg_loss = fluid.layers.mean(x=loss)
    fluid.backward.append_backward(avg_loss)
    return img, label, avg_loss


@ray.remote
class Worker(object):
    def __init__(self, worker_id):
        self.worker_id = worker_id
        self.scope = fluid.core.Scope()
        self.program = fluid.Program()
        self.startup = fluid.Program()
        with fluid.program_guard(self.program, self.startup):
            img, label, self.loss = mnist_network()

        self.place = fluid.CPUPlace()
        self.executor = fluid.Executor(self.place)
        self.reader_creator = paddle.batch(
            paddle.reader.shuffle(
                paddle.dataset.mnist.train(), buf_size=8192),
            batch_size=BATCH_SIZE)

        self.reader = self.reader_creator()
        self.feeder = fluid.DataFeeder(feed_list=[img, label], place=self.place)


    def compute_gradient(self, weights):
        for var_name in weights:
            tensor = self.scope.var(var_name).get_tensor()
            tensor.set(weights[var_name][0], self.place)
            tensor.set_lod(weights[var_name][1])

        try:
            data = next(self.reader)
        except:
            self.reader = self.reader_creator()
            data = next(self.reader)

        outs = self.executor.run(self.program,
                     feed=self.feeder.feed(data),
                               scope=self.scope,
                               fetch_list=[ var_name + "@GRAD" for var_name in weights] + [self.loss])
        if self.worker_id == 0:
            print outs[-1]
        return outs[:-1]


@ray.remote
class PServer(object):
    def __init__(self, learning_rate):
        self.scope = fluid.core.Scope()
        self.learning_rate = learning_rate
        self.program = fluid.Program()
        self.startup = fluid.Program()
        with fluid.program_guard(self.program, self.startup):
            mnist_network()

        self.place = fluid.CPUPlace()
        self.executor = fluid.Executor(self.place)
        self.executor.run(self.startup, scope=self.scope)
        self.optimize_program = fluid.Program()

    def apply_gradients(self, *gradients):
        # TODO(qijun) an optimization program is needed
        mean_gradients = numpy.mean(gradients, axis=0)
        weights = self.get_weight()
        for idx, name in enumerate(weights):
            w = weights[name][0]
            w -= self.learning_rate * mean_gradients[idx]
            self.scope.find_var(name).get_tensor().set(w, self.place)

    def get_weight(self):
        weights = dict()
        for p in self.program.global_block().iter_parameters():
            lod_tensor = self.scope.find_var(p.name).get_tensor()
            weights[p.name] = (numpy.copy(numpy.array(lod_tensor)), lod_tensor.lod())
        return weights


if __name__ == '__main__':
    ps = PServer.remote(1e-3 * NUM_WORKERS)
    weights = ps.get_weight.remote()

    works = [Worker.remote(i) for i in range(NUM_WORKERS)]

    while True:
        gradients = [work.compute_gradient.remote(weights) for work in works]
        ps.apply_gradients.remote(*gradients)
```

**Comment by typhoonzero** [comment]
@QiJune in the above code, `ps` should be a member of `Worker`, and the worker should call `get_weight` after `apply_gradients`.

And, yes, using `ray` be a simple way to implement dist train, but I doubt it could use more memory because it's using a `redis` instance to store `objects`, so all the weights in `pserver` must store both in fluid and redis so that workers can fetch them, and I must take a look into the implementaion of `ray` to confirm it that's true.

**Comment by reyoung** [comment]
> ps should be a member of Worker, and the worker should call get_weight after apply_gradients.

No. 

```
gradients = [work.compute_gradient.remote(weights) for work in works]
ps.apply_gradients.remote(*gradients)
```
This example contains N workers and one pservers. In the main function, these lines will wait all workers complete.

**Comment by typhoonzero** [comment]
> ps should be a member of Worker, and the worker should call get_weight after apply_gradients.

Sorry, my fault. `@ray.remote` decorated class is server instances, it should not be the member of the worker. Yet workers should update local weights after `ps.apply_gradients.remote(*gradients)`

**Comment by chengduoZH** [comment]
```
        try:
            data = next(self.reader)
        except:
            self.reader = self.reader_creator()
            data = next(self.reader)
```
This is to say that all the workers use their own `data reader`, right? @QiJune

**Comment by QiJune** [comment]
@chengduoZH This is just a simple demo following [sync_parameter_server](https://github.com/ray-project/ray/blob/master/examples/parameter_server/sync_parameter_server.py) in ray.
And we can add several data reader actors and a data provider actor for many works.

**Comment by QiJune** [comment]
@typhoonzero Has refine the code and  call `get_weight ` method after `apply_gradients `. The loss    curve seems normal.

**Comment by typhoonzero** [comment]
@QiJune Cool 👍!

**Comment by reyoung** [comment]
Just list some facts about `Ray`.

`Ray` might not be complete for our needs in several ways:

1. `Ray` is strongly tied to Python and using `pickle` to serialize objects.
2. The scheduler is implemented inside `Ray`. 
  * `Ray` can take over all CPUs and GPUs as a service, by `ray start`
  * In users' scripts, use `ray.init` can connect to the cluster service before. 
  * Actors are managed by `Ray` schedulers. i.e., which node will execute the actor is decided by `Ray` framework.
  * We cannot use `k8s` to explicitly customize the scheduler without digging into `Ray` deeply.
  * The scheduler of `Ray` is not mature for now(according the issues and documentation, there could be schedule two actors on the same GPU when there are actually two GPUs).
3. The `Ray` implements the `Actor Model`. However, `Ray` gives an object-oriented API for hiding `Actor Models`. Comparing the standard actor model, there are some [limits](http://ray.readthedocs.io/en/latest/actors.html#current-actor-limitations) of `Ray`.

Goods about `Ray`.

1. A separate project about communication.  It can be easily integrated into any framework.
2. Messages will only be read by its need. If an actor is just receiving a message and routing it to another actor, the message will not be read twice.
3. `Ray` implements `actor-model`, which is scalable and fault-tolerant by nature.
4. `Ray` has a [WebUI](http://ray.readthedocs.io/en/latest/webui.html) for debugging.

We may not use `Ray` directly. However, we can learn a lot from this project. We may implement a separate project for communication and provides C++ API to the fluid. Fluid may provide some operators or higher-level concepts to represent Actor Model or CSP. It seems to be a practical way.

**Comment by wangkuiyi** [comment]
I quickly went over Ray. It seems an implementation of the Actor model in Python? In my understanding, Ray is heavily integrated with Python. However, Fluid is trying to get out of Python. Are they having conflicting ideas?

**Comment by Yancey1989** [comment]
From @typhoonzero 
> for distributed training fluid API, we may have two main design:
Using channel
```python
def def_pserver():
    with fluid.listen_and_serv() as server:
        with fluid.While():
            grad = server.recv_chan.get()
            param = grad_param_mapping.get(grad.name)
            fluid.optimize(param, grad)
            server.return_chan.push(param)
```

For my personal opinion: 
1. Recv Op is executed in the background, so maybe we need to create a Goroutine and execute it and pass the channel as an argument.
1. A Channel need a type.

So, maybe the program would look like:

```python
def def_pserver():
    chan = fluid.Channel(type=sendrecv.message)
    with fluid.Go(chan):
        with fluild.listen_and_serv() as server:
                msg = server.get_msg()
                chan.push(msg)

    with fluid.While():
         grad = chan.get()
         param = grad_param_mapping.get(grad.name)
         fluid.optimize(param, grad)
         return_chan.push(param)
```

**Comment by shanyi15** [comment]
您好，此issue在近一个月内暂无更新，我们将于今天内关闭。若在关闭后您仍需跟进提问，可重新开启此问题，我们将在24小时内回复您。因关闭带来的不便我们深表歉意，请您谅解~感谢您对PaddlePaddle的支持! 
 Hello, this issue has not been updated in the past month. We will close it today for the sake of other user‘s experience. If you still need to follow up on this question after closing, please feel free to reopen it. In that case, we will get back to you within 24 hours. We apologize for the inconvenience caused by the closure and thank you so much for your support of PaddlePaddle Group!
